lib/rack/lint.rb

```diff
@@ -9,781 +9,864 @@ module Rack
+        raise LintError, "Response body must only be called once (#{@invoked})" unless @invoked.nil?
 
-      ##
-      ## If the Body responds to +to_path+, it must return a String
-      ## identifying the location of a file whose contents are identical
-      ## to that produced by calling +each+; this may be used by the
-      ## server as an alternative, possibly more efficient way to
-      ## transport the response.
+        @invoked = :each
 
-      if @body.respond_to?(:to_path)
-        unless ::File.exist? @body.to_path
-          raise LintError, "The file identified by body.to_path does not exist"
+        @body.each do |chunk|
+          ## and must only yield String values.
+          unless chunk.kind_of? String
+            raise LintError, "Body yielded non-string value #{chunk.inspect}"
+          end
+          @size += chunk.bytesize
+          yield chunk
+        end
+
+        verify_content_length(@size)
+
+        ##
+        ## If the Body responds to +to_path+, it must return a String
+        ## identifying the location of a file whose contents are identical
+        ## to that produced by calling +each+; this may be used by the
+        ## server as an alternative, possibly more efficient way to
+        ## transport the response.
+        ##
+        if @body.respond_to?(:to_path)
+          unless ::File.exist? @body.to_path
+            raise LintError, "The file identified by body.to_path does not exist"
+          end
         end
       end
 
       ##
-      ## The Body commonly is an Array of Strings, the application
-      ## instance itself, or a File-like object.
-    end
+      ## ==== Streaming Body
+      ##
+      def call(stream)
+        ## The Streaming Body must respond to +call+.
+        raise LintError, "Chunked Body must respond to call" unless @body.respond_to?(:call)
 
-    def close
-      @closed = true
-      @body.close  if @body.respond_to?(:close)
-    end
+        ## It must only be called once.
+        raise LintError, "Response body must only be called enumerated once (#{@invoked})" unless @invoked.nil?
 
-    # :startdoc:
+        @invoked = :call
 
+        ## It takes a +stream+ argument.
+        ##
+        ## The +stream+ argument must implement:
+        ## <tt>read, write, flush, close, close_read, close_write, closed?</tt>
+        ##
+        @body.call(StreamWrapper.new(stream))
+      end
+
+      class StreamWrapper
+        extend Forwardable
+
+        ## The semantics of these IO methods must be a best effort match to
+        ## those of a normal Ruby IO or Socket object, using standard arguments
+        ## and raising standard exceptions. Servers are encouraged to simply
+        ## pass on real IO objects, although it is recognized that this approach
+        ## is not directly compatible with HTTP2.
+        REQUIRED_METHODS = [
+          :read, :write, :flush, :close,
+          :close_read, :close_write, :closed?
+        ]
```

Comment:
I'm guessing there is a reason, but can you explain why `read` and `close_read` would be needed for response streaming, which should be write-only?  Is this due to TLS read-before-write issues?

Reply:
The provided stream can do both input and output, just like partial hijack.